torchdata/nodes/samplers/multi_node_weighted_sampler.py

```diff
@@ -105,31 +116,35 @@ def _validate(self) -> None:
 
     def reset(self, initial_state: Optional[Dict[str, Any]] = None):
         super().reset(initial_state)
-
         if initial_state is not None:
             self._num_yielded = initial_state[self.NUM_YIELDED_KEY]
-            self._weighted_sampler = _WeightedSampler(
-                weights=self.weights,
-                seed=self.seed,
-                rank=self.rank,
-                world_size=self.world_size,
-                initial_state=initial_state[self.WEIGHTED_SAMPLER_STATE_KEY],
-            )
+            self._epoch_random_seed = initial_state[self.EPOCH_RANDOM_SEED_KEY]
+            self._weighted_sampler = self._get_new_weighted_sampler(initial_state)
             self._datasets_exhausted = initial_state[self.DATASETS_EXHAUSTED_KEY]
             for k in self.dataset_names:
                 self.source_nodes[k].reset(initial_state[self.DATASET_NODE_STATES_KEY][k])
         else:
             # Force a fresh iterator from all source nodes
             self._num_yielded = 0
-            self._weighted_sampler = _WeightedSampler(
-                weights=self.weights,
-                seed=self.seed,
-                rank=self.rank,
-                world_size=self.world_size,
-            )
+
+            if self._started:
+                self._epoch_random_seed += 1
+                self._weighted_sampler = self._get_new_weighted_sampler()
```

Comment:
```suggestion
            if self._started:
                self._epoch_random_seed += 1
            self._weighted_sampler = self._get_new_weighted_sampler()
```
I think this works if we drop the call in `__init__`, it's a little easier to reason about